### ISSUE TITLE
batch>1 single-token decode returns corrupted logits (rows > 0) — mild at prefill, severe at decode; not model- or quantization-specific

### ISSUE THREAD

**Issue by TipKnuckle** [ml-explore/mlx-lm#1457]
> **Note on provenance.** This was surfaced by Claude (Anthropic's Fable
> model) while working with me on an experimental entropy-sampler repo,
> and the subject matter is honestly out of my personal depth. I've
> verified it as far as I'm able with agent assistance before filing. It
> reproduces on two unrelated architectures (Gemma-4 and Qwen3) and on
> **unquantized bf16** weights, so it's neither Gemma- nor
> quantization-specific. The underlying cause is a `mx.fast.rope`
> dispatch bug traced to a specific line, filed as
> **ml-explore/mlx#3795**; this issue is the mlx-lm-level symptom.
> Please correct anything I've mischaracterized.

**Repo:** ml-explore/mlx-lm

## Summary

In batched generation, logits from a batch>1 forward deviate from the
equivalent batch=1 computation. It reproduces on **different
architectures (Gemma-4 and Qwen3) and on unquantized bf16 weights**, so
it is neither a Gemma nor a quantization issue — it tracks input shape,
and points at the batched single-token-decode rope path (filed
separately, see root-cause section). Severity depends on shape:

- **Batched prefill (seq > 1): mild or clean.** Gemma dense 31B (4-bit),
  batch-2 with two distinct same-length prompts: max logit deviation
  0.59 / 1.38 vs each prompt alone; top-1 stable. Qwen3-4B (bf16) with
  identical prompts is exact (0.0). Consistent with ordinary
  kernel-switch numerics.
- **Batched single-token decode (seq == 1, with KV cache): severe, and
  the signature is identical across models** — row 0 exact/near-exact,
  every row ≥ 1 corrupted:
  - **Gemma dense 31B (4-bit):** 8 candidate tokens decoded in one
    batch-8 step against identical tiled caches, vs each decoded alone —
    row 0 max|d| = 0.000, **rows 1–7 deviate by 17–27.5 with top-1 flips
    on all seven**, after an 11-token prompt. At 5.6k context, same
    pattern (rows 1–7 deviating 26–33).
  - **Qwen3-4B (bf16, unquantized, dense):** same setup — row 0
    max|d| = 0.203 (bf16 noise floor), **rows 1–7 deviate by 6.4–14.8**,
    i.e. 30–70× the row-0 floor. Top-1 happened not to flip here (this
    model's top token was well-separated), but the logits are corrupted
    by the same order of magnitude and will flip tokens wherever the top
    two are closer, and corrupt any sampling. Being bf16, this rules out
    quantization as the cause.
- **Quantized MoE (gemma-4-26B-A4B-it-qat-4bit): catastrophic.**
  Batch-2 prefill with *identical* prompts: the two rows agree with
  each other but sit 6.94 from the batch=1 result; batched generation
  degenerates into word salad. (The extra prefill-shape deviation here,
  where dense models are clean, suggests a second MoE-specific component
  on top of the shared decode bug — see below.)

## Environment

- mlx-lm 0.31.3, mlx 0.31.2 (pip)
- macOS 26.5.1, Apple M4 Max, 128 GB
- Models: mlx-community/gemma-4-31B-it-qat-4bit,
  mlx-community/gemma-4-26B-A4B-it-qat-4bit,
  mlx-community/Qwen3-4B-bf16 (dense, unquantized)

## Repro 1 — severe decode case (dense 31B)

```python
import copy
import mlx.core as mx
from mlx_lm import load
from mlx_lm.models.cache import make_prompt_cache

model, tok = load("mlx-community/gemma-4-31B-it-qat-4bit")
ids = tok.encode("The old lighthouse keeper climbed the spiral stairs and")

cache = make_prompt_cache(model)
out = model(mx.array([ids]), cache=cache)
mx.eval(out)
cands = mx.argpartition(-out[0, -1], kth=7)[:8].tolist()

def tile(cache, k):
    tiled = []
    for c in cache:
        nc = copy.copy(c)
        if getattr(c, "keys", None) is not None:
            nc.keys = mx.repeat(c.keys, k, axis=0) if k > 1 else c.keys[:]
            nc.values = mx.repeat(c.values, k, axis=0) if k > 1 else c.values[:]
        tiled.append(nc)
    return tiled

gt = []
for t in cands:
    o = model(mx.array([[t]]), cache=tile(cache, 1))
    mx.eval(o)
    gt.append(o[0, -1])

ob = model(mx.array([[t] for t in cands]), cache=tile(cache, 8))
mx.eval(ob)
for i in range(8):
    d = mx.abs(ob[i, -1] - gt[i]).max().item()
    agree = mx.argmax(ob[i, -1]).item() == mx.argmax(gt[i]).item()
    print(f"row {i}: max|d| = {d:7.3f}   top-1 agree: {agree}")
```

Observed:

```
row 0: max|d| =   0.000   top-1 agree: True
row 1: max|d| =  21.875   top-1 agree: False
row 2: max|d| =  19.125   top-1 agree: False
row 3: max|d| =  19.625   top-1 agree: False
row 4: max|d| =  27.500   top-1 agree: False
row 5: max|d| =  17.250   top-1 agree: False
row 6: max|d| =  17.000   top-1 agree: False
row 7: max|d| =  21.250   top-1 agree: False
```

## Repro 2 — MoE case (26B)

```python
import mlx.core as mx
from mlx_lm import load

model, tok = load("mlx-community/gemma-4-26B-A4B-it-qat-4bit")
ids = tok.encode("The capital of France is")

b1 = model(mx.array([ids]))[0, -1]
b2 = model(mx.array([ids, ids]))
print("batch2 row0 vs row1 :", mx.abs(b2[0, -1] - b2[1, -1]).max().item())
print("batch2 row0 vs batch1:", mx.abs(b2[0, -1] - b1).max().item())
```

Observed:

```
batch2 row0 vs row1 : 0.0
batch2 row0 vs batch1: 6.9375
```

(Identical rows agreeing with each other while both being wrong is
expected if the corruption reads data across rows — identical rows make
cross-row reads invisible.)

## Repro 3 — non-Gemma, unquantized (Qwen3-4B bf16)

Repro 1 verbatim with `load("mlx-community/Qwen3-4B-bf16")` — a dense,
bf16, different-architecture model. Rules out both Gemma and
quantization as causes.

Observed:

```
== prefill, identical rows ==      # seq > 1: clean
  batch2 row0 vs row1 : 0.0
  batch2 row0 vs batch1: 0.0
== decode, 8 distinct tokens in one batch-8 step ==   # seq == 1
  row 0: max|d| =    0.203   top-1 agree: True
  row 1: max|d| =    8.312   top-1 agree: True
  row 2: max|d| =   11.875   top-1 agree: True
  row 3: max|d| =    6.438   top-1 agree: True
  row 4: max|d| =    7.438   top-1 agree: True
  row 5: max|d| =   13.562   top-1 agree: True
  row 6: max|d| =   10.562   top-1 agree: True
  row 7: max|d| =   14.750   top-1 agree: True
```

Same signature as Gemma (row 0 at the noise floor, rows ≥ 1 corrupted,
decode-only). Top-1 did not flip on this model/prompt because its top
token was well-separated; the logit corruption is the same order of
magnitude and flips tokens wherever the top two are closer.

## Root-cause isolation (all on the dense 31B)

- **QuantizedLinear is exonerated**: `[8, hidden]` matmuls with
  distinct rows match per-row results exactly at M ≤ 8 (small ~1e-2
  relative deviations only appear at M ≥ 16, ordinary kernel-switch
  noise).
- **Fused SDPA is exonerated for the decode case**: replacing it with
  explicit unfused attention did not change the deviations.
- **`mx.fast.rope` reproduces the signature exactly**: on contiguous
  `[B, H, 1, D]` input with B > 1 — precisely the batched-decode rope
  shape — row 0 is exact and rows ≥ 1 come back corrupted,
  dtype-independent. **Root cause filed and traced in ml-explore/mlx
  #3795** — the single-time-step fast-path in `rope.cpp` omits the batch
  dimension from its launch grid, so only batch row 0 is written and
  rows ≥ 1 return uninitialized memory. A model-free repro and the exact
  line are in that issue. The seq==1-only trigger also explains why
  batched *prefill* (seq > 1) is merely mild while batched *decode* is
  severe.
- The MoE 26B's additional prefill-shape deviation (6.94 at seq=6,
  where the dense model shows ≤ 1.4) suggests a second, MoE-specific
  component (quantized `gather_mm`?) on top of the rope issue.

## Expected

Batch>1 logits equal to the corresponding batch=1 computation within
ordinary kernel-dispatch noise (~1e-1 on these 4-bit models), no top-1
flips, coherent batched generation.

**Comment by TipKnuckle** [comment]
Closing — the root cause here is the `mx.fast.rope` batched single-token decode bug (mlx #3494), already fixed by ml-explore/mlx#3498 on `main`. My repro was on a release (0.31.2) that predates the fix, so this is resolved upstream, not an mlx-lm issue. Apologies for the duplicate.
